include/geopackage/NGen_SQLite.hpp

```diff
@@ -0,0 +1,236 @@
+
+  public:
+    sqlite_iter(stmt_t stmt);
+
+    /**
+     * Check if a row iterator is finished
+     * 
+     * @return true if next() returned SQLITE_DONE
+     * @return false if there is more rows available
+     */
+    bool done() const noexcept;
+
+    /**
+     * Step into the next row of a SQLite query
+     * 
+     * If the query is finished, next() acts idempotently,
+     * but will change done() to return true.
+     * @return sqlite_iter& returns itself
+     */
+    sqlite_iter& next();
+
+    /**
+     * Restart an iteration to its initial state.
+     * next() must be called after calling this.
+     * 
+     * @return sqlite_iter& returns itself
+     */
+    sqlite_iter& restart();
+
+    /**
+     * Call the row iterator destructor
+     */
+    void close();
+
+    /**
+     * Get the current row index for the iterator
+     * 
+     * @return int the current row index, or -1 if next() hasn't been called
+     */
+    int current_row() const noexcept;
+
+    /**
+     * Get the number of columns within this iterator
+     * @return int number of columns in query
+     */
+    int num_columns() const noexcept;
+
+    /**
+     * Return the column index for a named column 
+     * 
+     * @param name column name to search for
+     * @return int index of given column name, or -1 if not found.
+     */
+    int column_index(const std::string& name) const noexcept;
+
+    /**
+     * Get a vector of column names
+     * 
+     * @return const std::vector<std::string>& column names as a vector of strings
+     */
+    const std::vector<std::string>& columns() const noexcept { return this->column_names; }
+
+    /**
+     * Get a vector of column types
+     *
+     * See https://www.sqlite.org/datatype3.html for type values. The integers
+     * are the affinity for data types.
+     * @return const std::vector<int>& column types as a vector of ints
+     */
+    const std::vector<int>& types() const noexcept { return this->column_types; }
+
+    /**
+     * Get a column value from a row iterator by index
+     * 
+     * @tparam T Type to parse value as, i.e. int
+     * @param col Column index to parse
+     * @return T value at column `col`
+     */
+    template<typename T>
+    T get(int col) const;
+
+    /**
+     * Get a column value from a row iterator by name
+     * 
+     * @tparam T Type to parse value as, i.e. int
+     * @return T value at the named column
+     */
+    template<typename T>
+    T get(const std::string&) const;
+};
+
+template<typename T>
+inline T sqlite_iter::get(const std::string& name) const
+{
+    const int index = this->column_index(name);
+    return this->get<T>(index);
+}
+
+/**
+ * Wrapper around a SQLite3 database
+ */
+class sqlite
+{
+  private:
+    /**
+     * Smart pointer (unique) type for sqlite3 database
+     */
+    using sqlite_t = std::unique_ptr<sqlite3, sqlite_deleter>;
+  
+    sqlite_t conn = nullptr;
+    stmt_t   stmt = nullptr;
```

Comment:
Why does `class sqlite` have a `stmt` member at all? It doesn't do anything with it besides store it. Having it muddies ownership of the underlying `sqlite3_stmt` object, which could otherwise be kept in a `unique_ptr` in `sqlite_iter`

Reply:
Implemented changes in 914ec44: removed `stmt` member, changed `stmt_t` alias to a `unique_ptr`.